### PR TITLE
Add WarpX example for FEL simulation

### PR DESCRIPTION
**Pull request by RemiLehe**
This adds an example for how to run FEL simulations with the boosted-frame technique.
https://warpx--5337.org.readthedocs.build/en/5337/usage/examples/free_electron_laser/README.html

### CODE REVIEW

File: Examples/Physics_applications/free_electron_laser/README.rst [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+For `MPI-parallel <https://www.mpi-forum.org>`__ runs, prefix these lines with ``mpiexec -n 4 ...`` or ``srun -n 4 ...``, depending on the system.
+
+.. literalinclude:: inputs_test_1d_fel
+   :language: ini
+   :caption: You can copy this file from ``Examples/Physics_applications/beam_beam_collision/inputs_test_1d_fel``.
+
+
+Visualize
+---------
```
Comment (RemiLehe) [inline Examples/Physics_applications/free_electron_laser/README.rst]:
```suggestion
```

File: Examples/Physics_applications/free_electron_laser/analysis_fel.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+# Check that the radiation wavelength is the expected one
+iteration_check = 18
+Ex, info = ts_lab.get_field("E", "x", iteration=iteration_check)
+Nz = len(info.z)
+fft_E = abs(np.fft.fft(E_lab))
+lambd = 1.0 / np.fft.fftfreq(Nz, d=info.dz)
+lambda_radiation_lab = lambd[fft_E[:Nz].argmax()]
+lambda_expected = lambda_u / (2 * gamma_boost**2)
+# assert abs(lambda_radiation_lab - lambda_expected) / lambda_expected < 0.01
```
Comment (RemiLehe) [inline Examples/Physics_applications/free_electron_laser/analysis_fel.py]: This will be updated after https://github.com/ECP-WarpX/WarpX/pull/5226

**Comment by RemiLehe** [comment]
Close  #4861

File: Docs/source/usage/parameters.rst [hunk 3]
```diff
@@ -257,7 +257,7 @@ Overall simulation parameters
         ``warpx.self_fields_absolute_tolerance``).
 
     * ``fft``: Poisson's equation is solved using an Integrated Green Function method (which requires FFT calculations).
-        See these references for more details :cite:t:`QiangPhysRevSTAB2006`, :cite:t:`QiangPhysRevSTAB2006err`.
+        See these references for more details :cite:t:`param-QiangPhysRevSTAB2006`, :cite:t:`param-QiangPhysRevSTAB2006err`.
```
Comment (RemiLehe) [inline Docs/source/usage/parameters.rst:260]: This is somewhat orthogonal to this PR, but when building the Doc locally, I was getting warnings because of this.

File: Examples/Physics_applications/free_electron_laser/README.rst [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+Free-electron laser
+===================
+
+This example shows how to simulate the physics of a free-electron laser (FEL) using WarpX.
+In this example, a relativistic electron beam is sent through an undulator (represented by an external,
+oscillating magnetic field). The radiation emitted by the beam grows exponentially
+as the beam travels through the undulator, due to the Free-Electron-Laser instability.
+
+The parameters of the simulation are taken from section 5.1 of :cite:t:`ex-Fallahi2020`.
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/README.rst:11]: Why is there an `ex-` here and in the references below?

File: Examples/Physics_applications/free_electron_laser/README.rst [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+The simulation is performed in 1D, and uses the boosted-frame technique as described in
+:cite:t:`ex-VayFELA2009` and :cite:t:`ex-VayFELB2009`, to reduce the computational cost of the simulation.
+(The Lorentz frame of the simulation is moving at the average speed of the beam in the undulator.)
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/README.rst]: Could we make the parenthesis part of the sentence that it refers to here (if I understood correctly)?
```suggestion
The simulation is performed in 1D, and uses the boosted-frame technique as described in
:cite:t:`ex-VayFELA2009` and :cite:t:`ex-VayFELB2009` to reduce the computational cost (the Lorentz frame of the simulation is moving at the average speed of the beam in the undulator).
```

File: Examples/Physics_applications/free_electron_laser/README.rst [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+laboratory frame, using WarpX's ``BackTransformed`` diagnostic.
+
+The effect of space-charge is intentionally turned off in this example, as it may not be properly modeled in 1D.
+This is achieved by initializing two species of opposite charge (electrons and positrons) to
+represent the physical electron beam, as discussed in :cite:t:`ex-VayFELB2009`.
+
+Run
+---
+
+For `MPI-parallel <https://www.mpi-forum.org>`__ runs, prefix these lines with ``mpiexec -n 4 ...`` or ``srun -n 4 ...``, depending on the system.
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/README.rst]: It's not clear to me which "lines" (that should be prefixed with `mpiexec` or `srun`) this sentence is referring to. I see only the input file after this sentence.

File: Examples/Physics_applications/free_electron_laser/README.rst [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+.. literalinclude:: inputs_test_1d_fel
+   :language: ini
+   :caption: You can copy this file from ``Examples/Physics_applications/free_electron_laser/inputs_test_1d_fel``.
+
+Visualize
+---------
+
+The figure below shows the results of the simulation. The left panel shows the exponential growth of the radiation
+along the undulator. (Note that the vertical scale is in log scale.) The right panel shows a snapshot of the simulation,
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/README.rst]: Similarly to the comment above, could we make the parenthesis part of the sentence that it refers to here (if I understood correctly)?
```suggestion
The figure below shows the results of the simulation. The left panel shows the exponential growth of the radiation along the undulator (note that the vertical axis is plotted in log scale). The right panel shows a snapshot of the simulation,
```

File: Examples/Physics_applications/free_electron_laser/analysis_fel.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+import os
+import sys
+
+import numpy as np
+from openpmd_viewer import OpenPMDTimeSeries
+from scipy.constants import c, e, m_e
+
+sys.path.insert(1, "../../../../warpx/Regression/Checksum/")
+import checksumAPI
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/analysis_fel.py]: In #5297 we made all analysis scripts uniform with respect to the checksums parts. The new docs should reflect this (i.e., they should report code snippets that hopefully contributors will copy/paste in the future, keeping the uniformity). For now the uniformity is primarily aimed at making some follow-up changes coming up soon a bit easier to implement across the code base:
```suggestion
from checksumAPI import evaluate_checksum
```

File: Examples/Physics_applications/free_electron_laser/analysis_fel.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+lambd = 1.0 / np.fft.fftfreq(Nz, d=info.dz)
+lambda_radiation_boost = lambd[fft_E[: Nz // 2].argmax()]
+lambda_radiation_lab = lambda_radiation_boost / (2 * gamma_boost)
+lambda_expected = lambda_u / (2 * gamma_boost**2)
+assert abs(lambda_radiation_lab - lambda_expected) / lambda_expected < 0.01
+
+# Evaluate the checksum
+filename = sys.argv[1]
+test_name = os.path.split(os.getcwd())[1]
+checksumAPI.evaluate_checksum(test_name, filename, output_format="openpmd")
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/analysis_fel.py]: Same comment as above related to #5297:
```suggestion
# compare checksums
evaluate_checksum(
    test_name=os.path.split(os.getcwd())[1],
    output_file=sys.argv[1],
    output_format="openpmd",
)
```

File: Examples/Physics_applications/free_electron_laser/analysis_fel.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+k_u = 2 * np.pi / lambda_u
+K = e * Bu / (m_e * c * k_u)  # Undulator parameter
+gamma_boost = (
+    gamma_bunch / (1 + K * K / 2) ** 0.5
+)  # Lorentz factor of the ponderomotive frame
+beta_boost = (1 - 1.0 / gamma_boost**2) ** 0.5
+
+
+# Analyze the diagnostics showing quantities in the lab frame
+ts_lab = OpenPMDTimeSeries("diags/diag_labframe")
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/analysis_fel.py]: This could be read from the command line as well, right?
```suggestion
# Analyze the diagnostics showing quantities in the lab frame
filename = sys.argv[1]
ts_lab = OpenPMDTimeSeries(filename)
```

File: Examples/Physics_applications/free_electron_laser/analysis_fel.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+# (i.e. the growth of the power is exponential) and fit a line to extract the
+# gain length.
+i_start = 6
+i_end = 23
+# Perform linear fit
+p = np.polyfit(z_lab_peak[i_start:i_end], log_P_peak[i_start:i_end], 1)
+# Extract the gain length
+Lg = 1 / p[0]
+Lg_expected = 0.22  # Expected gain length from https://arxiv.org/pdf/2009.13645
+print("Gain length: ", Lg)
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/analysis_fel.py]: Just a personal preference for easier formatting, in case you like it too:
```suggestion
print(f"Gain length: {Lg}")
```

File: Examples/Physics_applications/free_electron_laser/analysis_fel.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+# Check that the radiation wavelength is the expected one
+iteration_check = 14
+Ex, info = ts_lab.get_field("E", "x", iteration=iteration_check)
+Nz = len(info.z)
+fft_E = abs(np.fft.fft(Ex))
+lambd = 1.0 / np.fft.fftfreq(Nz, d=info.dz)
+lambda_radiation_lab = lambd[fft_E[: Nz // 2].argmax()]
+lambda_expected = lambda_u / (2 * gamma_boost**2)
+print("lambda_radiation_lab", lambda_radiation_lab)
+print("lambda_expected", lambda_expected)
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/analysis_fel.py]: Just a personal preference for easier formatting, in case you like it too:
```suggestion
print(f"lambda_radiation_lab: {lambda_radiation_lab}")
print(f"lambda_expected: {lambda_expected}")
```

File: Examples/Physics_applications/free_electron_laser/analysis_fel.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+# (i.e. the growth of the power is exponential) and fit a line to extract the
+# gain length.
+i_start = 16
+i_end = 25
+# Perform linear fit
+p = np.polyfit(z_lab_peak[i_start:i_end], log_P_peak[i_start:i_end], 1)
+# Extract the gain length
+Lg = 1 / p[0]
+Lg_expected = 0.22  # Expected gain length from https://arxiv.org/pdf/2009.13645
+print("Gain length: ", Lg)
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/analysis_fel.py]: Just a personal preference for easier formatting, in case you like it too:
```suggestion
print(f"Gain length: {Lg}")
```

File: Examples/Physics_applications/free_electron_laser/plot_sim.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
+import matplotlib.pyplot as plt
+from openpmd_viewer import OpenPMDTimeSeries
+
+ts = OpenPMDTimeSeries("./diags/diag_labframe/")
```
Comment (EZoni) [inline Examples/Physics_applications/free_electron_laser/plot_sim.py:4]: Just wondering if we should pass this file path as a command line argument as well and read it through `sys.argv[1]`, like in the analysis script, in case someone runs this manually outside of CTest and moves files around while doing so...

**Review by EZoni — CHANGES_REQUESTED** [review]
This is really well done, thank you! I left only minor comments.

File: Examples/Physics_applications/free_electron_laser/analysis_fel.py (hunk 11 above)
Comment (EZoni) [reply Examples/Physics_applications/free_electron_laser/analysis_fel.py]: P.S. f-strings make multi-parameter formatting (and spacing) a bit easier in my opinion, e.g.
```py
print(f"Gain length between {i_start} and {i_end}: {Lg}")
```

File: Examples/Physics_applications/free_electron_laser/README.rst (hunk 4 above)
Comment (RemiLehe) [reply Examples/Physics_applications/free_electron_laser/README.rst:11]: Hm, I am not quite sure. I followed the same citation format as here:
https://github.com/ECP-WarpX/WarpX/blob/development/Examples/Physics_applications/laser_ion/README.rst

File: Examples/Physics_applications/free_electron_laser/README.rst (hunk 6 above)
Comment (RemiLehe) [inline Examples/Physics_applications/free_electron_laser/README.rst]:
```suggestion
This example can be run with the WarpX executable using an input file: ``warpx.1d inputs_test_1d_fel``. For `MPI-parallel <https://www.mpi-forum.org>`__ runs, prefix these lines with ``mpiexec -n 4 ...`` or ``srun -n 4 ...``, depending on the system.
```

Comment (RemiLehe) [reply Examples/Physics_applications/free_electron_laser/README.rst]: Thanks for catching this. There is indeed something missing.

File: Examples/Physics_applications/free_electron_laser/plot_sim.py (hunk 14 above)
Comment (RemiLehe) [reply Examples/Physics_applications/free_electron_laser/plot_sim.py:4]: I am OK leaving it as is for now, but then changing the `plot.py` in this example as well as other examples in a follow PR, if we think this is how users should run it.

File: Examples/Physics_applications/free_electron_laser/README.rst (hunk 4 above)
Comment (EZoni) [reply Examples/Physics_applications/free_electron_laser/README.rst:11]: Okay, I will check separately what that means and whether it's the right way for citing references with reST.

**Review by EZoni — APPROVED** [review]
